### DISCUSSION TITLE
Checking against null with is null does not work

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#5376] [Q&A]
I have already reported this problem which already has been identified by other and seems to be a compiler bug, but I cannot find my original post.

Originally I'm checking reference types by comparing it to null

`if (ReferenceType == null) throw ArgumentNullException();`

Now, the analyzers suggest, I should to do 

`if (ReferenceType is null) throw ArgumentNullException();`

But this seems to be a problem in some cases and falls through the condition even if the ReferenceType is null. Finding these problems is really annoying and it would really be helpful if this problem is solved.

Maybe someone could link me to the original finding of this problem.

**Comment by svick** [comment]
Are you talking about https://github.com/dotnet/roslyn/issues/45030?

**Reply by msedi** [reply]
@svick: No, it was a different thread. Unfortunately I can't remember if I opened the issue or someone else.

**Comment by Joe4evr** [comment]
Does your `ReferenceType` have an overloaded `==` operator that special-cases certain states as equivalent to `null` (something that Unity's `GameObject` is(/was?) also known for)? If yes, then the refactoring suggestion here should be ignored, because the pattern-syntax translates to `((object)ReferenceType) == null` which bypasses that overloaded operator.

This is not new, and the language team have always said that refactoring suggestions are not guaranteed to behave identical to what the code does prior, specifically because of cases like this.

Now, could the analyzer detect that the type has an overloaded `==` operator in it's hierarchy and not show the suggestion for those cases? Absolutely. But probably >99% of those implementations will do comparison with "real" `null`s for equality, making the behavior after the refactor equivalent to prior, so the suggestion wouldn't be given in a majority of cases where it is valid.

Now, if you *don't* have an overloaded `==` (or don't special-case), then I have no idea. ¯\\\_(ツ)_/¯

**Comment by msedi** [comment]
@Joe4evr. I agree with you with the overloaded behavior. On one object I indeed have an overloaded operator, but on the other I do not have any == overloads. 

As stated above I'm trying to find the issue and come back.

**Comment by msedi** [comment]
Found it: #4433
